File: src/main/old/core/Table.java
package core;

import deck.*;
import piles.*;

import java.util.*;

import static core.Constants.WHITE_TEXT;

public class Table {
    private DrawPile drawPile;
    private UncoverdPile uncoveredPile;

    private Map<Suit, SuitPile> suitPiles = new HashMap<>();

    private Lane[] lanes = new Lane[7];
    public Table(){
        Deck deck = new Deck();
        for(int i=0;i<7;i++){
            lanes[i] = new Lane(deck.getCards(i+1));
        }
        for(Lane lane : lanes){
            lane.getTopCard().flip();
        }
        for(Suit suit : Suit.values()){
            suitPiles.put(suit,new SuitPile(suit));
        }
        uncoveredPile = new UncoverdPile();
        drawPile = new DrawPile(deck.getCards(deck.getDeckSize()));
    }

    public void printTable(){
        int lines=0;
        //find the longest lane
        for(Lane lane : lanes){
            if(lane.getSize() > lines)
                lines=lane.getSize();
        }
        //print labels
        System.out.println(WHITE_TEXT +"     P \t     C   D   H   S");
        System.out.print(drawPile + " " + uncoveredPile + "\t    ");
        for(Suit suit : Suit.values()){
            System.out.print(suitPiles.get(suit)+" ");
        }
        System.out.print("\n\n ");
        //print lane labels
        for(int i=0;i < lanes.length;i++){
            System.out.print(WHITE_TEXT +(i+1)+"   ");
        }
        System.out.println();
        //print lanes
        for(int i=0;i<lines;i++){
            for (Lane lane : lanes){
                Card card = lane.getCard(i);
                if(card!=null){
                    System.out.print(card + " ");
                }
                else{
                    System.out.print("    ");
                }
            }
            System.out.println();
        }
    }

    public static void moveCard(Pile from, Pile to) throws InvalidMoveException {
        Card card = from.getTopCard();
        to.addCard(card);
        from.removeCard(card);
        if(from.getTopCard() != null && from.getTopCard().isCovered() && !(from instanceof DrawPile))
            from.flipTopCard();
    }

    public static void moveCards(int number, Pile from, Pile to) throws InvalidMoveException {
        if (number==0 || from.isEmpty()){
            throw new InvalidMoveException();
        }
        List<Card> cards=new LinkedList<>();
        for(int i= from.getSize()-number;i < from.getSize();i++){
            if(from.getCard(i)!=null)
                cards.add(from.getCard(i));
        }
        to.addCards(cards);
        from.removeCards(cards);
        if(from.getTopCard() != null && from.getTopCard().isCovered())
            from.flipTopCard();
    }

    public static void moveCards(Pile from, Pile to) throws InvalidMoveException {
        List<Card> cards=new LinkedList<>();
        for(int i= from.getSize()-1;i > 0;i--){
            if(from.getCard(i)!=null)
                cards.add(from.getCard(i));
        }
        to.addCards(cards);
        from.removeCards(cards);
        if(from.getTopCard() != null && from.getTopCard().isCovered())
            from.flipTopCard();
    }

    public void drawCard() throws InvalidMoveException {
        if(drawPile.isEmpty()){
            if(uncoveredPile.isEmpty()){
                throw new InvalidMoveException();
            }
            moveCards(uncoveredPile,drawPile);
            for(Card card : drawPile){
                card.flip();
            }
        }
        else{
            moveCard(drawPile,uncoveredPile);
        }
    }

    public Lane getLane(int laneNumber) {
        return lanes[laneNumber];
    }

    public Pile getSuitPile(Suit suit) {
        return suitPiles.get(suit);
    }

    public UncoverdPile getUncoveredPile() {
        return uncoveredPile;
    }
}

File: app/src/main/scala/core/Table.scala
package core

import core.Constants.WHITE_TEXT
import deck.*
import piles.{DrawPile, Lane, Pile, SuitPile, UncoveredPile}

import scala.Console.out


object Table {
  @throws[InvalidMoveException]
  def moveCard(from: Pile, to: Pile): Unit = {
    val card = from.getTopCard.get
    to.addCard(card)
    from.removeCard(card)
    from.getTopCard match
      case Some(topCard) => from match
        case _:DrawPile =>
        case _ => if (topCard.covered) from.flipTopCard()
      case None =>
  }

  @throws[InvalidMoveException]
  def moveCards(number: Int, from: Pile, to: Pile): Unit = {
    if (number == 0 || from.isEmpty) throw new InvalidMoveException
    var cards = Vector[Card]()
    for (i <- (from.getSize - number) until from.getSize){
      from.getCard(i) match
        case Some(x) => cards = cards :+ x
        case None => 
    }
    to.addCards(cards)
    from.removeCards(cards)
    from.getTopCard match
      case Some(topCard) => if (topCard.covered) from.flipTopCard()
      case None =>
  }

  @throws[InvalidMoveException]
  def moveCards(from: Pile, to: Pile): Unit = {
    var cards = Vector[Card]()
    for (i <- (from.getSize - 1) until 0 by -1) {
      from.getCard(i) match
        case Some(c) => cards = cards :+ c
        case None =>
    }
    to.addCards(cards)
    from.removeCards(cards)
    from.getTopCard match
      case Some(topCard) => if (topCard.covered) from.flipTopCard()
      case None =>
  }
}

class Table {
  val deck = new Deck
  private var suitPiles: Map[Suit, SuitPile] = Map[Suit,SuitPile]()
  private var lanes: Seq[Lane] = Vector[Lane]()
  for (i <- 1 to 7) {
    lanes = lanes :+ new Lane(deck.getCards(i))
  }
  for (lane <- lanes) {
    lane.getTopCard.get.flip()
  }

  for (suit <- Suit.values) {
    suitPiles = suitPiles + (suit -> new SuitPile(suit))
  }

  private var drawPile: DrawPile = new DrawPile(deck.getCards(deck.getDeckSize))
  var uncoveredPile: UncoveredPile = new UncoveredPile

  def printTable(): Unit = {
    var lines = 0
    //find the longest lane
    for (lane <- lanes) {
      if (lane.getSize > lines) lines = lane.getSize
    }
    //print labels
    out.println(WHITE_TEXT + "     P \t     C   D   H   S")
    out.print("" + drawPile + " " + uncoveredPile + "\t    ")
    for (suit <- Suit.values) {
      out.print("" +
        (suitPiles.get(suit) match
        case Some(x) => x
        case None => "   " )
        + " ")
    }
    out.print("\n\n ")
    //print lane labels
    for (i <- lanes.indices) {
      out.print(WHITE_TEXT + (i + 1) + "   ")
    }
    out.println()
    //print lanes
    for (i <- 0 until lines) {
      for (lane <- lanes) {
        out.print("" +
          (lane.getCard(i) match
          case Some(c) => c
          case None => "   ")
            + " ")
      }
      out.println()
    }
  }

  @throws[InvalidMoveException]
  def drawCard(): Unit = {
    if (drawPile.isEmpty) {
      if (uncoveredPile.isEmpty) throw new InvalidMoveException
      Table.moveCards(uncoveredPile, drawPile)
      for (card <- drawPile) {
        card.flip()
      }
    }
    else Table.moveCard(drawPile, uncoveredPile)
  }

  def getLane(laneNumber: Int): Lane = lanes(laneNumber)

  def getSuitPile(suit: Suit): SuitPile = suitPiles(suit)
}
